`AbstractInterpreter`: define `add_invalidation_callback!` utility

aviatesk opened the pull request:
It seems we're reaching a consensus on how the external `AbstractInterpreter` utilizes user invalidation callbacks. Let's establish a utility within `Core.Compiler` for this purpose, which cannibalizes the pattern, so that external packages can easily access and use it.

aviatesk committed:
`AbstractInterpreter`: define `add_invalidation_callback!` utility

It seems we're reaching a consensus on how the external
`AbstractInterpreter` utilizes user invalidation callbacks.
Let's establish a utility within `Core.Compiler` for this purpose,
which cannibalizes the pattern, so that external packages can easily
access and use it.

vchuravy commented:
This seems fine, but I was thinking the other day if we could avoid the callback... I think it's one of the pieces that blocks us from having across session caching.

Instead we could tag the MI (or even the world) with information about the AbsInt it is valid in

aviatesk commented on base/compiler/utilities.jl:
```diff
@@ -323,6 +323,44 @@ function iterate(iter::BackedgeIterator, i::Int=1)
+                                          seen::IdSet{MethodInstance}=IdSet{MethodInstance}())
+    push!(seen, replaced)
+    callback.callback(replaced)
+    if isdefined(replaced, :backedges)
```
Hmm, how about invoking external `callbacks` directly from `invalidate_backedges`? This would allow external callbacks to avoid traversing backedges themselves. I've looked through `gf.c` but didn't see any reasons why we shouldn't call `invalidate_external` from within `invalidate_backedges`.

aviatesk replied:
Something like:
```diff
diff --git a/base/compiler/utilities.jl b/base/compiler/utilities.jl
index ca513db952..ec2e6dd8ad 100644
--- a/base/compiler/utilities.jl
+++ b/base/compiler/utilities.jl
@@ -327,13 +327,10 @@ end
     add_invalidation_callback!(callback, mi::MethodInstance)
 
 Register `callback` to be triggered upon the invalidation of `mi`.
-`callback` should a function taking a single argument, `replaced::MethodInstance`,
+`callback` should a function taking two arguments, `callback(replaced::MethodInstance, max_world::UInt32)`,
 and it will be recursively invoked on `MethodInstance`s within the invalidation graph.
 """
-add_invalidation_callback!(@nospecialize(callback), mi::MethodInstance) =
-    _add_invalidation_callback!(mi, InvalidationCallback(callback))
-
-function _add_invalidation_callback!(mi::MethodInstance, @nospecialize(callback))
+function add_invalidation_callback!(@nospecialize(callback), mi::MethodInstance)
     if !isdefined(mi, :callbacks)
         callbacks = mi.callbacks = Any[callback]
     else
@@ -345,22 +342,6 @@ function _add_invalidation_callback!(mi::MethodInstance, @nospecialize(callback)
     return callbacks
 end
 
-struct InvalidationCallback{Callback}
-    callback::Callback
-end
-function (callback::InvalidationCallback)(replaced::MethodInstance, max_world,
-                                          seen::IdSet{MethodInstance}=IdSet{MethodInstance}())
-    push!(seen, replaced)
-    callback.callback(replaced)
-    if isdefined(replaced, :backedges)
-        for mi in replaced.backedges::Vector{Any}
-            isa(mi, MethodInstance) || continue # might be `Type` object representing an `invoke` signature
-            mi in seen && continue # otherwise fail into an infinite loop
-            callback(mi, max_world, seen)
-        end
-    end
-end
-
 #########
 # types #
 #########
diff --git a/src/gf.c b/src/gf.c
index 407d9c517a..3debbd15f0 100644
--- a/src/gf.c
+++ b/src/gf.c
@@ -1639,6 +1639,8 @@ static void invalidate_external(jl_method_instance_t *mi, size_t max_world) {
 
 static void do_nothing_with_codeinst(jl_code_instance_t *ci) {}
 
+static void _invalidate_backedges(void (*f)(jl_code_instance_t*), jl_method_instance_t *replaced_mi, size_t max_world, int depth);
+
 // recursively invalidate cached methods that had an edge to a replaced method
 static void invalidate_method_instance(void (*f)(jl_code_instance_t*), jl_method_instance_t *replaced, size_t max_world, int depth)
 {
@@ -1667,28 +1669,14 @@ static void invalidate_method_instance(void (*f)(jl_code_instance_t*), jl_method
         JL_GC_POP();
         codeinst = jl_atomic_load_relaxed(&codeinst->next);
     }
+    invalidate_external(replaced, max_world);
     // recurse to all backedges to update their valid range also
-    jl_array_t *backedges = replaced->backedges;
-    if (backedges) {
-        JL_GC_PUSH1(&backedges);
-        replaced->backedges = NULL;
-        size_t i = 0, l = jl_array_len(backedges);
-        jl_method_instance_t *replaced;
-        while (i < l) {
-            i = get_next_edge(backedges, i, NULL, &replaced);
-            invalidate_method_instance(f, replaced, max_world, depth + 1);
-        }
-        JL_GC_POP();
-    }
+    _invalidate_backedges(f, replaced, max_world, depth + 1);
     JL_UNLOCK(&replaced->def.method->writelock);
 }
 
-// invalidate cached methods that overlap this definition
-static void invalidate_backedges(void (*f)(jl_code_instance_t*), jl_method_instance_t *replaced_mi, size_t max_world, const char *why)
-{
-    JL_LOCK(&replaced_mi->def.method->writelock);
+static void _invalidate_backedges(void (*f)(jl_code_instance_t*), jl_method_instance_t *replaced_mi, size_t max_world, int depth) {
     jl_array_t *backedges = replaced_mi->backedges;
-    //jl_static_show(JL_STDERR, (jl_value_t*)replaced_mi);
     if (backedges) {
         // invalidate callers (if any)
         replaced_mi->backedges = NULL;
@@ -1701,6 +1689,13 @@ static void invalidate_backedges(void (*f)(jl_code_instance_t*), jl_method_insta
         }
         JL_GC_POP();
     }
+}
+
+// invalidate cached methods that overlap this definition
+static void invalidate_backedges(void (*f)(jl_code_instance_t*), jl_method_instance_t *replaced_mi, size_t max_world, const char *why)
+{
+    JL_LOCK(&replaced_mi->def.method->writelock);
+    _invalidate_backedges(f, replaced_mi, max_world, 1);
     JL_UNLOCK(&replaced_mi->def.method->writelock);
     if (why && _jl_debug_method_invalidation) {
         jl_array_ptr_1d_push(_jl_debug_method_invalidation, (jl_value_t*)replaced_mi);
@@ -1814,6 +1809,7 @@ static int invalidate_mt_cache(jl_typemap_entry_t *oldentry, void *closure0)
             }
         }
         if (intersects) {
+            // TODO call invalidate_external here?
             if (_jl_debug_method_invalidation) {
                 jl_array_ptr_1d_push(_jl_debug_method_invalidation, (jl_value_t*)mi);
                 jl_value_t *loctag = jl_cstr_to_string("invalidate_mt_cache");
@@ -2060,7 +2056,6 @@ JL_DLLEXPORT void jl_method_table_insert(jl_methtable_t *mt, jl_method_t *method
                 }
                 if (missing) {
                     jl_method_instance_t *backedge = (jl_method_instance_t*)backedges[i];
-                    invalidate_external(backedge, max_world);
                     invalidate_method_instance(&do_nothing_with_codeinst, backedge, max_world, 0);
                     invalidated = 1;
                     if (_jl_debug_method_invalidation)
diff --git a/test/compiler/EscapeAnalysis/EAUtils.jl b/test/compiler/EscapeAnalysis/EAUtils.jl
index e19f6fdfae..5b4e58c056 100644
--- a/test/compiler/EscapeAnalysis/EAUtils.jl
+++ b/test/compiler/EscapeAnalysis/EAUtils.jl
@@ -137,7 +137,7 @@ CC.getindex(wvc::WorldView{EscapeAnalyzerCacheView}, mi::MethodInstance) = getin
 function CC.setindex!(wvc::WorldView{EscapeAnalyzerCacheView}, ci::CodeInstance, mi::MethodInstance)
     wvc.cache.code_cache.cache[mi] = ci
     # register the callback on invalidation
-    CC.add_invalidation_callback!(mi) do replaced::MethodInstance
+    CC.add_invalidation_callback!(mi) do replaced::MethodInstance, max_world
         delete!(wvc.cache.code_cache.cache, replaced)
         delete!(wvc.cache.escape_cache.cache, replaced)
     end
diff --git a/test/compiler/invalidation.jl b/test/compiler/invalidation.jl
index 96b61503de..fe7e2f9683 100644
--- a/test/compiler/invalidation.jl
+++ b/test/compiler/invalidation.jl
@@ -46,9 +46,9 @@ CC.get(wvc::WorldView{InvalidationTesterCacheView}, mi::MethodInstance, default)
 CC.getindex(wvc::WorldView{InvalidationTesterCacheView}, mi::MethodInstance) = getindex(wvc.cache.dict, mi)
 CC.haskey(wvc::WorldView{InvalidationTesterCacheView}, mi::MethodInstance) = haskey(wvc.cache.dict, mi)
 function CC.setindex!(wvc::WorldView{InvalidationTesterCacheView}, ci::CodeInstance, mi::MethodInstance)
-    CC.add_invalidation_callback!(mi) do replaced::MethodInstance
+    CC.add_invalidation_callback!(mi) do replaced::MethodInstance, max_world
         delete!(wvc.cache.dict, replaced)
-        # Core.println(replaced) # debug
+        Core.println("[InvalidationTester] ", replaced) # debug
     end
     setindex!(wvc.cache.dict, ci, mi)
 end
```

aviatesk replied:
What do you think about c97093f? That makes user invalidation callback kicked on each method instance invalidation, allowing it not to traverse its backedge list by itself.

/cc @vchuravy @maleadt With this change, we will no longer need to traverse backedge list in external packages, although we may still need to tweak `max_world` of invalidated `CodeInstance`s (, which is already done within GPUCompiler).

vchuravy replied:
I am okay with that change, but I also haven't thought about this too hard.


One thing I would encourage you to think about is how this interact with Cross-Session caching. That's a problem that we need to tackle sooner than later since AbstractInterpreter are becoming a significant contributor to latency across the ecosystem.

vtjnash replied:
it is just a bad idea for external consumers to be relying on this unstable and internal datastructure at all. The user can implement their own clone of it, but the user really shouldn't not be trying to poke at the internal one directly.

aviatesk replied:
Certainly, I agree that `MethodInstance` is an internal data structure and is inherently unstable. Yet, given the present `AbstractInterpreter` interface, `MethodInstance` remains crucial for inference, making it as a caching key feels intuitive, and so banning its use might pose significant challenges (though I agree with your long-term viewpoint). As for this PR, this eliminates necessity for users to access `mi.backedges`, which I feel strictly improves the current status. Thus, I'd like to move forward with it as is.

aviatesk committed:
call `invalidate_external` from `invalidate_method_instance`

So that user invalidation callback does not need to traverse backedges by themselves.
